gdb/mips-tdep.c

```diff
@@ -757,6 +757,33 @@ mips_pseudo_register_read (struct gdbarch *gdbarch, struct regcache *regcache,
   int rawnum = cookednum % gdbarch_num_regs (gdbarch);
   gdb_assert (cookednum >= gdbarch_num_regs (gdbarch)
 	      && cookednum < 2 * gdbarch_num_regs (gdbarch));
+  /*
+   * For hybrid binaries in a sandbox we also need to consider $pcc.base when
+   * computing the program counter: https://github.com/CTSRD-CHERI/gdb/issues/1
+   */
+  int cap_pcc = gdbarch_tdep (gdbarch)->regnum->cap_pcc;
+  if (rawnum == gdbarch_tdep (gdbarch)->regnum->pc && cap_pcc >= 0)
+  {
+      /*
```

Comment:
I think it would be better to not duplicate the code but use the existing get_cheri_register_signed() function in this file.  However, I don't know that this is really going to do what you want for unwinding since at some point you will want stack frames to start reading PC again.  That is, if you do 'up' out of the sandbox and into plain MIPS this might try to use PCC which would be from the first sandbox frame.  The real answer is that the compiler needs to emit proper DWARF info that lets us know when to switch ABIs for hybrid binaries.  Without that it's actually pretty hard to make hybrid work correctly.  To date my focus has been on getting a pure CheriABI binary to work and I believe those should work with sandboxes now as-is.

Reply:
(Ironically, GDB's notion of the ABI to use for stack traces, etc. is tied to each shared object, so if we were to use dynamic linking where sandbox code was always in a separate .so that was tagged with EF_MIPS_CHERIABI, then the existing logic should work fine for hybrid binaries by using PCC in the sandbox and falling back to plain MIPS when unwinding out)

Reply:
Currently, there is no useful backtrace anyway since GDB doesn't know about the files that libcheri has mapped. But at least it gives me the correct program counter value in the backtrace.

Ideally we would somehow tell libcheri to do the r_debug stuff that rtld does but I don't quite understand how that would work.

Reply:
So true sandbox debugging will be quite a bit more involved.  We will need a way to enumerate sandboxes similar to how we use r_debug to enumerate shared libraries.  We will have to adjust symbol resolution to take the current compartment into account (kind of like TLS variables).  This gets more interesting if we use a shared read-only PT_LOAD (.text and .rodata) across compartments but separate read-write PT_LOAD (.data and .bss) per compartment in that rtld's existing table will probably cover the read-only PT_LOAD's already, and the read-write ones are only needed to handle symbol resolution for per-compartment variables.  I think right now libcheri isn't doing anything like that.  As long as we have a lm_info like thing in libcheri that is probably enough to come up with a modified solib hook to enumerate the files libcheri maps.  It's somewhat messy as we still want the existing solib-svr4 logic for rtld itself.  In some ways it would be a lot nicer if libcheri would maybe just call into rtld to add files it maps to rtld's existing linked list?  Eventually I think we want that part of libcheri in rtld anyway and if we do that then the existing solib handling would work.

Reply:
Yes, I think having libcheri as part of rtld will make things a lot easier.

libcheri currently does have a list of mapped files but I'm not sure what needs to be done in order for GDB to be able to parse it.